cluster-autoscaler/proposals/plugable-provider-grpc.md

```diff
@@ -0,0 +1,604 @@
+the current cloud provider implementations.
+
+## Goals
+
+* Support custom cloud provider implementations without changing the current `CloudProvider` interface.
+* Make CA extendable, so users do not need to fork the CA repository.
+
+## Proposal
+
+There are couple of examples of pluggable designs using Go SDKs that would guide us on how to extend CA
+to support custom providers as plugins.
+This approach is inspired based on [Hashicorp go-plugin](https://github.com/hashicorp/go-plugin) and [Grafana Go SDK for plugins](https://github.com/grafana/grafana-plugin-sdk-go).
+
+The proposed solution will deploy the external grpc provider along side CA and both communicates via gRPC with TLS/SSL. The external provider would be part of a separate deployment, and so we should deploy it independently, as shown in the ![Figure](./images/external-provider-grpc.jpg).
+
+This approach exposes a common gRPC API server.
+
+
+
+The API calls would be defined by the operations exposed by the `CloudProvider` and `NodeGroup` interfaces in [CA](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/cloud_provider.go#L50):
+
+```go
+type CloudProvider interface {
+	Name() string
+
+	NodeGroups() []NodeGroup
+
+	NodeGroupForNode(*apiv1.Node) (NodeGroup, error)
+
+	Pricing() (PricingModel, errors.AutoscalerError)
+
+	NewNodeGroup(machineType string, labels map[string]string, systemLabels map[string]string,
+		taints []apiv1.Taint, extraResources map[string]resource.Quantity) (NodeGroup, error)
+
+	GetResourceLimiter() (*ResourceLimiter, error)
+
+	GPULabel() string
+
+	GetAvailableGPUTypes() map[string]struct{}
+
+	Cleanup() error
+
+	Refresh() error
+
+  ...
+}
+
+type NodeGroup interface {
+	MaxSize() int
+
+	MinSize() int
+
+	TargetSize() (int, error)
+
+	IncreaseSize(delta int) error
+
+	DeleteNodes([]*apiv1.Node) error
+
+	DecreaseTargetSize(delta int) error
+
+	Id() string
+
+	Nodes() ([]Instance, error)
+
+	TemplateNodeInfo() (*schedulerframework.NodeInfo, error)
+
+	Exist() bool
+
+	Create() (NodeGroup, error)
+
+	Delete() error
+
+	Autoprovisioned() bool
+}
+
+```
+
+In the following I detail the main operations of the CloudProvider GRPC service:
+
+```protobuf
+syntax = "proto3";
+
+package clusterautoscaler.cloudprovider.v1;
+
+import "google/protobuf/descriptor.proto";
+import "google/protobuf/timestamp.proto";
+import "google/protobuf/wrappers.proto";
+
+import "k8s.io/apimachinery/pkg/apis/meta/v1/generated.proto";
+import "k8s.io/apimachinery/pkg/api/resource/generated.proto";
+import "k8s.io/api/core/v1/generated.proto"
+
+option go_package = "clusterautoscaler.cloudprovider";
+
+service CloudProvider {
+  // CloudProvider specific RPC functions
+  rpc Name(NameRequest)
+    returns (NameResponse) {}
+
+  rpc NodeGroups(NodeGroupsRequest)
+    returns (GetNameResponse) {}    
+
+  rpc NodeGroupForNode(NodeGroupForNodeRequest)
+    returns (NodeGroupForNodeResponse) {}
+
+  rpc PricingNodePrice(PricingNodePriceRequest)
```

Comment:
This is only needed for price expander and most existing providers don't implement it. I don't mind having it, but we should at least explicitly say implementing this is optional (at least beyond implementing an empty stub response).

Reply:
I added `Optional` descriptions in the operation and the operation definition.